### DISCUSSION TITLE
Caching of negative responses if the returned query type is different from queried type.

### DISCUSSION THREAD

**Discussion by alpominth** [0xERR0R/blocky#832] [Ideas]
When using **dig** to query a domain that has no support from queried type, a SOA response is gotten:

```
$ time dig @127.0.0.1 -p 53 -t PTR t.co

; <<>> DiG 9.18.4-2ubuntu2-Ubuntu <<>> @127.0.0.1 -p 53 -t PTR t.co
; (1 server found)
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 33069
;; flags: qr rd ra; QUERY: 1, ANSWER: 0, AUTHORITY: 1, ADDITIONAL: 1

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 4096
;; QUESTION SECTION:
;t.co.                          IN      PTR

;; AUTHORITY SECTION:
t.co.                   28      IN      SOA     ns1.p26.dynect.net. ops.twitter.com. 2538 3600 600 604800 60

;; Query time: 233 msec
;; SERVER: 127.0.0.1#53(127.0.0.1) (UDP)
;; WHEN: Tue Jan 17 09:13:01 EST 2023
;; MSG SIZE  rcvd: 102


real    0m0.252s
user    0m0.003s
sys     0m0.005s
```

![212575431-45dc80e0-4fcd-4094-8971-a15a5212921e](https://user-images.githubusercontent.com/122581130/212920420-7b3eb0d9-5c6a-4827-9dcd-19311805ddbe.jpg)

Blocky cache this negative answer, but it doesn't cache the returned type (SOA) when querying the cache:

```
$ time dig @127.0.0.1 -p 53 -t PTR t.co

; <<>> DiG 9.18.4-2ubuntu2-Ubuntu <<>> @127.0.0.1 -p 53 -t PTR t.co
; (1 server found)
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 59575
;; flags: qr rd ra; QUERY: 1, ANSWER: 0, AUTHORITY: 0, ADDITIONAL: 0

;; QUESTION SECTION:
;t.co.                          IN      PTR

;; Query time: 0 msec
;; SERVER: 127.0.0.1#53(127.0.0.1) (UDP)
;; WHEN: Tue Jan 17 09:13:47 EST 2023
;; MSG SIZE  rcvd: 22


real    0m0.019s
user    0m0.002s
sys     0m0.006s
```

I think Blocky should cache the returned query type even if it is different from the queried type.

**Comment by alpominth** [comment]
@0xERR0R @ThinkChaos 

Is this behavior intended in Blocky?

If it's not, could you please modify to cache the returned queried type if it's different from queried one?

**Reply by 0xERR0R** [reply]
Blocky caches only the "answer" section. In your case the result has no answer. If your query for t.co with "-t SOA", you'll get the answer section. I'm not sure if we should cache also the authority section. Some DNS servers will return it and some not.

**Reply by alpominth** [reply]
I tested with almost all DNS forwarders I know (dnscrypt-proxy, dnsproxy, overture, SmartDNS, Glider, Unbound, Knot Resolver) and all of them cache the authority section. And as I could see, all the DNS software written in Go which uses the package [miekg/dns](https://github.com/miekg/dns) and that cache DNS queries, all of them cache the authority section.

I could use any other DNS forwarder with cache, but what makes me using Blocky is the Redis cache feature together with separated caching of negative responses.

My recommendation is that you implement the caching of authority section.

**Reply by 0xERR0R** [reply]
@kwitsch @ThinkChaos any thoughts?

**Reply by alpominth** [reply]
At least it would be one more feature for Blocky, I see as an advantage.

**Reply by kwitsch** [reply]
Seams reasonable, no objections from my side.

**Reply by 0xERR0R** [reply]
I created https://github.com/0xERR0R/blocky/issues/833 to track it

**Reply by ThinkChaos** [reply]
Yeah seems reasonable like this. There's definitely some caching improvements still to do!

**Reply by alpominth** [reply]
I'm looking forward for a commit. =)

**Reply by alpominth** [reply]
@0xERR0R @kwitsch @ThinkChaos

Is there an expected date for this be implemented? Maybe you could address this issue on 0.21 milestone.

Sorry for the bother, but I really need this feature.
